### DISCUSSION TITLE
Where can I find the block findInst implementation?

### DISCUSSION THREAD

**Discussion by oharboe** [The-OpenROAD-Project/OpenROAD#3871] [General]
Silly question: where do I find the source code for `findInst`?

I seem unable to find it in the OpenROAD source code.

```
    set inst [$block findInst [format "ces_%d_%d" $row $col]]
```

https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/blob/f90d853b3705338fa2d142acd20ac57626c71f12/flow/designs/asap7/mock-array/macro-placement.tcl#L10

**Comment by QuantamHD** [comment]
Header: https://github.com/The-OpenROAD-Project/OpenROAD/blob/master/src/odb/include/odb/db.h#L992
Impl: https://github.com/The-OpenROAD-Project/OpenROAD/blob/master/src/odb/src/db/dbBlock.cpp#L1790
Hidden Impl: https://github.com/The-OpenROAD-Project/OpenROAD/blob/master/src/odb/src/db/dbBlock.h#L173

**Reply by oharboe** [reply]
Ah, I see. I expected to see something in .tcl that interfaced to this code, but I guess tcl can then call C++ methods directly without them being in the .i files. Not quite sure what those .i files are, but most things I call from .tcl seems to be in OpenROAD .i files...

**Reply by QuantamHD** [reply]
The swig'd part comes from this line https://github.com/The-OpenROAD-Project/OpenROAD/blob/master/src/odb/src/swig/common/odb.i#L85

**Comment by oharboe** [comment]
So `findInst` doesn't throw an exception if the element doesn't exist, like https://cplusplus.com/reference/map/map/at/ does.

This leads to a lot of repetition and context unique error messages and in tcl very poor error messages like:

```
Error: macro-placement.tcl, 27 invalid command name "NULL"
```

`std::map` has explicit accessors that either creates an element if it doesn't exist(returns empty by default) and another that throws an exception if it doesn't exist.

The equivalent here would be that findInst() threw an exception with an error message and the name, and that another hasInst() that doesn't and allows the user to handle it. try/catch is not approperiate because both findInst() that throws exception and hasInst() that checks for presence would both be part of normal exception.

**Reply by QuantamHD** [reply]
I think we could potentially wrap this routine in an exception in the TCL/SWIG layer, but I would caution against modifying the C++.

We have to use those routines inside the engine itself, and exception code has a non-zero cost.

**Comment by maliberty** [comment]
The odb code is old and was written without the use of exceptions.  I would be fine to change that but it will require testing the full app to see if problems arise and resolving them.  There are a number of similar APIs that return nullptr on failure.

**Comment by oharboe** [comment]
I see. I was just curious and surprised: there are LOT of methods in that .h file that are exposed in .tcl.

Not all of them are obviously compatible with a .tcl calling convention or even intended for .tcl?

**Reply by maliberty** [reply]
Returning NULL can be checked for in TCL.  Some that take reference arguments are trickier but there is some swig magic to help with many of those (at least the ones we've had occasion to need).

**Reply by oharboe** [reply]
Makes sense.

I was thinking that there could be a new getInst() that throws an exception with a readable error message if there's no instance. Invocations that do not check findInst() return value or that manually check and throw an exception, could be changed to getInst().

I searched for findInst() and there must be a dozen or more error numbers allocated to findInst() returning null + tcl code would be simplified and cleaned up.

**Reply by maliberty** [reply]
Feel free to make things better

**Reply by oharboe** [reply]
Seems like we are agreed that adding getInst() sibling to findInst() would be an improvement then.

I don't want to work on it unless there's a consensus of what is better, but once there is a consensus I might fix a few of the eyesores...

**Reply by maliberty** [reply]
I haven't inspected to see but I wonder if the caller who checks the return of findInst is in a position to give a better error message than getInst (ie some context about the lookup).

**Reply by oharboe** [reply]
That is exceedingly rare when one looks at exceptions generally. You will be hard pressed to find examples in codebases that use exceptions. It is not worth the complication, generally.

There can be cases where it is worth it. Such error messages should have corresponding test code...

**Reply by QuantamHD** [reply]
I'm personally not in favor of the C++ API returning exceptions. 

I think checking return types is generally pretty accepted. In python for example get None back is a totally expected thing, and works nicely with the C++.

Exceptions are also a non-zero runtime overhead in arguably a pretty hot code area.

**Reply by oharboe** [reply]
Exceptions are used pervasively in the OpenROAD codebase, so nothing new there 

In the proposal there is an exception and another exception free interface, also found pervasively in recent exception code.

**Reply by maliberty** [reply]
I think of exceptions as something that went wrong in the code.  If a user gives a name that isn't in the db that isn't exceptional.  A getter throwing an exception implies that something was definitely expected to be found and its absence represents something unforeseen.

**Reply by oharboe** [reply]
Indeed. "find" vs "get" reflects the expectations with the caller.

**Reply by maliberty** [reply]
Your motivating statement
```
I searched for findInst() and there must be a dozen or more error numbers allocated to findInst() returning null + tcl code would be simplified and cleaned up.
```
isn't clearly correct.  Many of those may be findInst and not getInst.

**Reply by QuantamHD** [reply]
It seems to me that this would be better handled by a TCL wapper script. Just use findInst to implement getInst

**Reply by oharboe** [reply]
46 hits, if I believe vscode :-)

Checking a few... sorting them into what I think was the intent of the caller.

getInst() case:

https://github.com/The-OpenROAD-Project/OpenROAD/blob/ad1deef3d5f598a812f66e42e05ff87c51aff05b/src/cts/src/TritonCTS.cpp#L630

possibly this one: https://github.com/The-OpenROAD-Project/OpenROAD/blob/ad1deef3d5f598a812f66e42e05ff87c51aff05b/src/odb/src/db/dbSearch.cpp#L1553



findInst() case:

https://github.com/The-OpenROAD-Project/OpenROAD/blob/ad1deef3d5f598a812f66e42e05ff87c51aff05b/src/dbSta/src/dbNetwork.cc#L479

unclear:

https://github.com/The-OpenROAD-Project/OpenROAD/blob/ad1deef3d5f598a812f66e42e05ff87c51aff05b/src/dbSta/src/dbNetwork.cc#L502
https://github.com/The-OpenROAD-Project/OpenROAD/blob/ad1deef3d5f598a812f66e42e05ff87c51aff05b/src/dbSta/src/dbNetwork.cc#L472

Should this one have a more general mechanism for logging exceptions and continuing? Continuing after an exception, something truely unexpected and unsupported is a dubious policy. https://github.com/The-OpenROAD-Project/OpenROAD/blob/ad1deef3d5f598a812f66e42e05ff87c51aff05b/src/odb/src/defin/definBlockage.cpp#L82

**Reply by maliberty** [reply]
cts : shouldn't be using names at all - should have stored the object when created
dbSearch : is a warning and represents a find operation
dbNetwork : all are implementing find
defin: intentionally warns and errors at the end.  This allows you to see all parsing problems instead of just the first.

None of these are good examples for getInst.

**Reply by oharboe** [reply]
Makes sense. I am leaving this be until the next crossroads where there is a motivation to look at it again.
